core/chapters/c06_lists.py

```diff
@@ -420,95 +420,277 @@ class index_error(Step):
+`0` is the default start, so `range(4)` is equal to `range(0, 4)`.
+`4` is the end of the range, but the end is always excluded, so the last value is `3`.
+If you're confused now, don't worry about it.
+
+There's a good reason for why `range(4)` is not actually a list - it makes programs faster and more efficient.
+It's not worth explaining that more right now.
+
+But you can easily convert it to a list: try `__program__` in the shell.
+
+What do you expect the output to be?
+        """
+        predicted_output_choices = [
+            "range(4)",
+            "range(0, 4)",
+            "list(range(4))",
+            "list(range(0, 4))",
+            "range(0, 1, 2, 3)",
+            "(0, 1, 2, 3)",
+            "[0, 1, 2, 3]"
+        ]
+
+        expected_code_source = "shell"
+
+        program = "list(range(4))"
+
+    class introducing_len_and_range(VerbatimStep):
+        """
+That's just a demonstration to let you see a range in a more familiar form.
+You should almost never actually do that.
+
+If you're feeling overwhelmed, don't worry! All you need to know is that `range(n)`
+is very similar to the list:
+
+    [0, 1, 2, ..., n - 2, n - 1]
+
+By the way, you can get the number of elements in a list (commonly called the *length*) using the `len` function.
+Try it by running this code:
+
+    __copyable__
+    __program_indented__
+        """
+
         auto_translate_program = False
 
+        predicted_output_choices = ["0", "1", "2", "3", "4"]
+
         def program(self):
             words = ['This', 'is', 'a', 'list']
+            print(len(words))
+
+    class print_last_element(ExerciseStep):
+        """
+Exercise: for any non-empty list `words`, print the last element. For example, if
+
+    __no_auto_translate__
+    words = ['This', 'is', 'a', 'list']
+
+your program should print `list`.
+        """
+
+        hints = """
+To access the last element of the list, you'll need the index of the last position.
+If the list has 2 elements, the first element is at index 0, so the last element is at index 1.
+Likewise, if the list had 3 elements, the last element would be at index 2.
+Do you see a pattern between those numbers? How can you express it?
+Can you come up with a general solution that works for any length?
+        """
+
+        def solution(self, words: List[str]):
+            print(words[len(words) - 1])
+
+        tests = [
+            (["Python"], "Python"),
+            (['Hello', 'world'], "world"),
+            (['futurecoder', 'is', 'cool!'], "cool!"),
+            (['This', 'is', 'a', 'list'], "list")
+        ]
+
+    class range_len(ExerciseStep):
+        """
+So in general, the valid indices are:
+
+    [0, 1, 2, ..., len(words) - 2, len(words) - 1]
+
+Exercise: for any non-empty list `words`, print each index and element of the list. For example, if
 
+    __no_auto_translate__
+    words = ['This', 'is', 'a', 'list']
+
+your program should print:
```

Comment:
```suggestion
No we can fix the program from earlier to work with any list. Fill in the `...`:

    __copyable__
    __no_auto_translate__
    words = ['This', 'is', 'a', 'list']

    for index in ...:
        print(index)
        print(words[index])

For the given example value of `words` it should print: 
```

Reply:
changed, I guess I should change the hints too?

Reply:
any suggestions?

Reply:
yeah you should probably change them, no suggestions from me. i don't think they need to be perfect, they should solve this easily.